### ISSUE TITLE
Horizontal tabs should be allowed in header values

### ISSUE THREAD

**Issue by MihaZupan** [dotnet/aspnetcore#40599]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

According to RFC, valid characters for a header value are visible ASCII characters (VCHAR), SP and HTAB.

https://datatracker.ietf.org/doc/html/rfc7230#section-3.2
```
header-field   = field-name ":" OWS field-value OWS
field-value    = *( field-content / obs-fold )
field-content  = field-vchar [ 1*( SP / HTAB ) field-vchar ]
field-vchar    = VCHAR / obs-text

VCHAR          = %x21-7E visible (printing) characters (https://datatracker.ietf.org/doc/html/rfc5234#appendix-B.1)
```

Kestrel allows VCHAR and SP, but prohibits HTAB.
https://github.com/dotnet/aspnetcore/blob/84c8d0f93cc4b6fbce1367ffc366a10037ecee77/src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpCharacters.cs#L110

IIS similarly excludes the whole `< 0x20` range.
https://github.com/dotnet/aspnetcore/blob/5a0e56e2dd05ae2aa7edde085aec8823b120d73a/src/Shared/HttpSys/RequestProcessing/HeaderCollection.cs#L277

Is this just an oversight, or was ignoring tabs a deliberate choice?

### Expected Behavior

HTAB (0x09) characters are allowed the same way as spaces.

### Steps To Reproduce

```c#
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

app.Map("/", (HttpContext context) =>
{
    context.Response.Headers.Add("X-Custom-Header", "Foo\tBar");
    return "Hello world";
});

app.Run();
```

### Exceptions (if any)

```
System.InvalidOperationException: Invalid non-ASCII or control character in header: 0x0009
    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.ThrowInvalidHeaderCharacter(Char ch)
    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.ValidateHeaderValueCharacters(String headerCharacters, Boolean requireAscii)
    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.ValidateHeaderValueCharacters(String headerName, StringValues headerValues, Func`2 encodingSelector)
    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpResponseHeaders.AddValueFast(String key, StringValues value)
    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.System.Collections.Generic.IDictionary<System.String,Microsoft.Extensions.Primitives.StringValues>.Add(String key, StringValues value)
```

### .NET Version

6.0 / 7.0 nightlies (not version dependant though)

### Anything else?

Hit by an internal partner using YARP with Kestrel on .NET 6.0.

I don't know of any workaround to avoid the validation of `< 0x20`, is there any?
While the HTAB can probably be safely added to the list of allowed characters, the team may wish to send control characters as well.

**Comment by MihaZupan** [comment]
cc: @suwatch @Tratcher @adityamandaleeka

**Comment by adityamandaleeka** [comment]
Reopening for servicing.

**Comment by adityamandaleeka** [comment]
Fixed in main and 6.0. Closing.
